### ISSUE TITLE
Sharing memory pools in CUDAGraphs has no effect on memory usage

### ISSUE THREAD

**Issue by ali-alshaar7** [Lightning-AI/lightning-thunder#1792]
*Note*: If you have a model or program that is not supported yet but should be, please use the program coverage template.

## 🐛 Bug

In theory, [sharing memory pools across CUDAGraphs](https://github.com/Lightning-AI/lightning-thunder/pull/1790) should decrease allocated memory, but in our experiments, it doesn't.

### To Reproduce

follow the test in the above PR.

### Expected behavior

The following minimal repro using pure torch cg does show difference between setting men_pool = None and men_pool = g1.pool()

```
import torch
def workload(a):
    inter = [a * i for i in range(50)]
    b = inter.pop()
    while inter:
        b = b + inter.pop()
    return b

g1 = torch.cuda.CUDAGraph()
g2 = torch.cuda.CUDAGraph()

n = 1024
m = 512

# (create static inputs for g1 and g2, run warmups of their workloads...)
static_in_1 = torch.randn(n, n, device="cuda")
static_in_2 = torch.randn(m, m, device="cuda")

static_out_1 = torch.randn_like(static_in_1)
static_out_2 = torch.randn_like(static_in_2)

for _ in range(3):
    with torch.no_grad():
        static_out_1.copy_(workload(static_in_1))
        static_out_2.copy_(workload(static_in_2))

# Captures g1
with torch.cuda.graph(g1):
    static_out_1.copy_(workload(static_in_1))

mem_pool = g1.pool()

# Captures g2, hinting that g2 may share a memory pool with g1
with torch.cuda.graph(g2, pool=mem_pool): #, pool=g1.pool()):
    static_out_2.copy_(workload(static_in_2))

real_data_1 = torch.randn_like(static_in_1)
real_data_2 = torch.randn_like(static_in_2)

static_in_1.copy_(real_data_1)
static_in_2.copy_(real_data_2)
g1.replay()
g2.replay()

from pprint import pprint

for seg in torch.cuda.memory_snapshot():
    pprint(seg)

reserved_memory_with_pool = torch.cuda.memory_reserved()
print(f"Memory reserved with pool: {reserved_memory_with_pool}")

```

### Environment

- PyTorch Version (e.g., 1.0):
- OS (e.g., Linux):
- How you installed PyTorch (`conda`, `pip`, source):
- Build command you used (if compiling from source):
- Python version:
- CUDA/cuDNN version:
- GPU models and configuration: T4 GPU on Ligtning Studio
- Any other relevant information:

**Comment by t-vi** [comment]
I think we have a more detailed picture now.

**Comment by csarofeen** [comment]
I took your simple script @ali-alshaar7 worked with @syed-ahmed and @eqy to make sure we could get a working example based off it, and instrumented it like crazy. expandable_segments is important though not necessary for demonstration. Feel free to play around with the two options. --use-memory-pool will do the shared pool with cuda graphs, and I put in --batch-size so you could easily play with another dimension.

I'd recommend playing around with this script, it does look like it's working as intended. Play around with --batch-size and should be able to see a higher limit with the --use-memory-pool option than without it.

You can also play around with the shapes to increase or decrease the second size. I just put them the same because it makes it very clear additional memory isn't used with the shared memory pool, but is without it.

CC @galv as we're talking about CUDA Graphs live tomorrow.

```.py
import os
os.environ["PYTORCH_CUDA_ALLOC_CONF"] = "expandable_segments:True"

import torch
import litgpt
from torch.testing import make_tensor
import argparse
device = torch.device("cuda:0")

# Parse command line arguments
parser = argparse.ArgumentParser()
parser.add_argument('--use-memory-pool', action='store_true', 
                   help='Use memory pool sharing between CUDA graphs')
parser.add_argument('--batch-size', type=int, default=1,
                   help='Batch size for input tensors')
args = parser.parse_args()

def print_memory_usage_info(msg):
    torch.cuda.synchronize()
    print(msg)
    print("Memory reserved:", round(torch.cuda.memory_reserved() / 1e6), "MB")
    print("Max memory reserved:", round(torch.cuda.max_memory_reserved() / 1e6), "MB")

def round_mb(size):
    return round(size / 1e6)

def print_memory_pool_info(pool_id):
    all_segments = torch.cuda.memory_snapshot()
    cudagraph_segments = [segment for segment in all_segments if segment["segment_pool_id"] == pool_id]
    print(f"Pool id: {pool_id} has {len(cudagraph_segments)} segments")
    for idx, segment in enumerate(cudagraph_segments):
        print(f"\t segment {idx}, type: {segment['segment_type']}, pool id: {segment['segment_pool_id']}, total_size: {round_mb(segment['total_size'])} MB, allocated_size: {round_mb(segment['allocated_size'])} MB, active_size: {round_mb(segment['active_size'])} MB, requested_size: {round_mb(segment['requested_size'])} MB")
    print()


cfg = litgpt.Config.from_name("open_llama_3b", n_layer=10)
print("Block size: ", cfg.block_size)
shape = (args.batch_size,) + (cfg.block_size,)
shape2 = (args.batch_size,) + (cfg.block_size,)

x = torch.randint(0, 255, shape, device=device, dtype=torch.long)
y = torch.randint(0, 255, shape2, device=device, dtype=torch.long)
model = litgpt.GPT(cfg).to(device)

print_memory_usage_info("Initial memory:")

g1 = torch.cuda.CUDAGraph()
g2 = torch.cuda.CUDAGraph()


torch.cuda.memory._record_memory_history()

for _ in range(3):
    with torch.no_grad():
        static_out_1 = model(x)

print_memory_usage_info("After initial model runs of first size:")

for _ in range(3):
    with torch.no_grad():
        static_out_2 = model(y)

print(static_out_2.dtype)
print(static_out_2.numel()/(1024**2), "MB")

print_memory_usage_info("After initial model runs of second size:")

with torch.cuda.graph(g1):
    with torch.no_grad():
        static_out_1.copy_(model(x))

print_memory_usage_info("After capturing first graph:")
print_memory_pool_info(g1.pool())

mem_pool = g1.pool() if args.use_memory_pool else None

with torch.cuda.graph(g2, pool=mem_pool):
    with torch.no_grad():
        static_out_2.copy_(model(y))

print_memory_usage_info("After capturing second graph:")

# Replay the graphs 5 times each
for _ in range(5):
    g1.replay()

print_memory_usage_info("After replaying graphs:")

for _ in range(5):
    g2.replay()

print_memory_usage_info("After replaying graphs:")

print_memory_pool_info(g1.pool())
print_memory_pool_info(g2.pool())
```

**Comment by ali-alshaar7** [comment]
> I took your simple script [@ali-alshaar7](https://github.com/ali-alshaar7) worked with [@syed-ahmed](https://github.com/syed-ahmed) and [@eqy](https://github.com/eqy) to make sure we could get a working example based off it, and instrumented it like crazy. expandable_segments is important though not necessary for demonstration. Feel free to play around with the two options. --use-memory-pool will do the shared pool with cuda graphs, and I put in --batch-size so you could easily play with another dimension.
> 
> I'd recommend playing around with this script, it does look like it's working as intended. Play around with --batch-size and should be able to see a higher limit with the --use-memory-pool option than without it.
> 
> You can also play around with the shapes to increase or decrease the second size. I just put them the same because it makes it very clear additional memory isn't used with the shared memory pool, but is without it.
> 
> CC [@galv](https://github.com/galv) as we're talking about CUDA Graphs live tomorrow.
> 
> import os
> os.environ["PYTORCH_CUDA_ALLOC_CONF"] = "expandable_segments:True"
> 
> import torch
> import litgpt
> from torch.testing import make_tensor
> import argparse
> device = torch.device("cuda:0")
> 
> # Parse command line arguments
> parser = argparse.ArgumentParser()
> parser.add_argument('--use-memory-pool', action='store_true', 
>                    help='Use memory pool sharing between CUDA graphs')
> parser.add_argument('--batch-size', type=int, default=1,
>                    help='Batch size for input tensors')
> args = parser.parse_args()
> 
> def print_memory_usage_info(msg):
>     torch.cuda.synchronize()
>     print(msg)
>     print("Memory reserved:", round(torch.cuda.memory_reserved() / 1e6), "MB")
>     print("Max memory reserved:", round(torch.cuda.max_memory_reserved() / 1e6), "MB")
> 
> def round_mb(size):
>     return round(size / 1e6)
> 
> def print_memory_pool_info(pool_id):
>     all_segments = torch.cuda.memory_snapshot()
>     cudagraph_segments = [segment for segment in all_segments if segment["segment_pool_id"] == pool_id]
>     print(f"Pool id: {pool_id} has {len(cudagraph_segments)} segments")
>     for idx, segment in enumerate(cudagraph_segments):
>         print(f"\t segment {idx}, type: {segment['segment_type']}, pool id: {segment['segment_pool_id']}, total_size: {round_mb(segment['total_size'])} MB, allocated_size: {round_mb(segment['allocated_size'])} MB, active_size: {round_mb(segment['active_size'])} MB, requested_size: {round_mb(segment['requested_size'])} MB")
>     print()
> 
> 
> cfg = litgpt.Config.from_name("open_llama_3b", n_layer=10)
> print("Block size: ", cfg.block_size)
> shape = (args.batch_size,) + (cfg.block_size,)
> shape2 = (args.batch_size,) + (cfg.block_size,)
> 
> x = torch.randint(0, 255, shape, device=device, dtype=torch.long)
> y = torch.randint(0, 255, shape2, device=device, dtype=torch.long)
> model = litgpt.GPT(cfg).to(device)
> 
> print_memory_usage_info("Initial memory:")
> 
> g1 = torch.cuda.CUDAGraph()
> g2 = torch.cuda.CUDAGraph()
> 
> 
> torch.cuda.memory._record_memory_history()
> 
> for _ in range(3):
>     with torch.no_grad():
>         static_out_1 = model(x)
> 
> print_memory_usage_info("After initial model runs of first size:")
> 
> for _ in range(3):
>     with torch.no_grad():
>         static_out_2 = model(y)
> 
> print(static_out_2.dtype)
> print(static_out_2.numel()/(1024**2), "MB")
> 
> print_memory_usage_info("After initial model runs of second size:")
> 
> with torch.cuda.graph(g1):
>     with torch.no_grad():
>         static_out_1.copy_(model(x))
> 
> print_memory_usage_info("After capturing first graph:")
> print_memory_pool_info(g1.pool())
> 
> mem_pool = g1.pool() if args.use_memory_pool else None
> 
> with torch.cuda.graph(g2, pool=mem_pool):
>     with torch.no_grad():
>         static_out_2.copy_(model(y))
> 
> print_memory_usage_info("After capturing second graph:")
> 
> # Replay the graphs 5 times each
> for _ in range(5):
>     g1.replay()
> 
> print_memory_usage_info("After replaying graphs:")
> 
> for _ in range(5):
>     g2.replay()
> 
> print_memory_usage_info("After replaying graphs:")
> 
> print_memory_pool_info(g1.pool())
> print_memory_pool_info(g2.pool())

So had a deeper look at the example and found that sharing memory pools is only effective if all the warmups are done prior to any graph captures, i.e
```
for _ in range(3):
    with torch.no_grad():
        static_out_1 = model(x)

print_memory_usage_info("After initial model runs of first size:")

for _ in range(3):
    with torch.no_grad():
        static_out_2 = model(y)

# capture g1
# capture g2
```
will save memory with pools, while

```
for _ in range(3):
    with torch.no_grad():
        static_out_1 = model(x)

print_memory_usage_info("After initial model runs of first size:")

# capture g1

for _ in range(3):
    with torch.no_grad():
        static_out_2 = model(y)
# capture g2
```
will not see any saving with pools.

`CUDAGraphsTranform` follows the second pattern since each input gets its own graph runner object which manages warmup, static in/out allocation and graph capture. We should look into allowing the first pattern to be used when we know what inputs we expect to use on the captured graph a priori.
